File: backend/app/ranking/scorer.py
"""Pontuacao de custo-beneficio 0-100, transparente e explicavel (secao 19).

Cada componente gera nota 0-100 + explicacao; a nota final e a soma ponderada
pelos pesos de configuracao (config.DEFAULT_WEIGHTS, ajustaveis por ambiente).
O prazo de entrega tem peso zero, mas gera alerta quando extremo/indefinido.
"""

import math

from app.core.config import get_settings
from app.schemas.models import HistoryStats, Offer, PriceBreakdown, RankedOffer, Reputation

_REPUTATION_SCORE = {
    "excelente": 100.0, "boa": 80.0, "regular": 55.0, "ruim": 25.0,
    "critica": 10.0, "insuficiente": 40.0, "nao_localizada": 40.0,
}
_HISTORY_SCORE = {
    "muito_baixo": 100.0, "baixo": 80.0, "na_media": 60.0,
    "alto": 30.0, "muito_alto": 10.0, "insuficiente": 50.0,
}
_REPUTATION_LABEL = {
    "excelente": "excelente", "boa": "boa", "regular": "regular", "ruim": "ruim",
    "critica": "crítica", "insuficiente": "insuficiente", "nao_localizada": "não localizada",
}


def _price_score(total: float, best_total: float) -> tuple[float, str]:
    if total <= 0 or best_total <= 0:
        return 0.0, "Preço total indisponível."
    ratio = best_total / total
    score = round(min(100.0, ratio * 100), 1)
    if total == best_total:
        return score, "Menor preço total entregue entre as ofertas válidas."
    pct = (total / best_total - 1) * 100
    return score, f"Preço total {pct:.0f}% acima da oferta mais barata válida."


def _reviews_score(offer: Offer) -> tuple[float, str]:
    r = offer.reviews
    if r.count == 0:
        return 30.0, "Sem avaliações localizadas; nota neutra baixa."
    confidence = min(1.0, math.log10(r.count + 1) / 3)  # ~1.0 com 1000+ avaliações
    base = r.average / 5 * 100
    score = round(base * (0.5 + 0.5 * confidence), 1)
    return score, (
        f"Nota {r.average:.1f}/5 em {r.count} avaliações "
        f"(confiança {'alta' if confidence > 0.8 else 'média' if confidence > 0.5 else 'baixa'})."
    )


def _reputation_score(store: Reputation, seller: Reputation) -> tuple[float, str]:
    s_store = _REPUTATION_SCORE[store.classification]
    if seller.classification in ("nao_localizada", "insuficiente"):
        score, txt = s_store, f"Loja com reputação {_REPUTATION_LABEL[store.classification]}."
    else:
        score = round(s_store * 0.6 + _REPUTATION_SCORE[seller.classification] * 0.4, 1)
        txt = (
            f"Loja {_REPUTATION_LABEL[store.classification]}, "
            f"vendedor {_REPUTATION_LABEL[seller.classification]}."
        )
    return score, txt


def _specs_score(desirable_met: list[str], total_desirable: int) -> tuple[float, str]:
    if total_desirable == 0:
        return 60.0, "Nenhum critério desejável informado; nota neutra."
    frac = len(desirable_met) / total_desirable
    score = round(40 + frac * 60, 1)  # atender tudo = 100; nada = 40
    return score, f"Atende {len(desirable_met)} de {total_desirable} critérios desejáveis."


def _history_score(history: HistoryStats) -> tuple[float, str]:
    score = _HISTORY_SCORE[history.classification]
    if history.classification == "insuficiente":
        return score, "Histórico de preços insuficiente; nota neutra."
    label = history.classification.replace("_", " ")
    return score, f"Preço atual classificado como '{label}' frente ao histórico."


def _warranty_score(offer: Offer) -> tuple[float, str]:
    w = offer.warranty
    if w.kind == "nacional":
        return (100.0, f"Garantia nacional de {w.months} meses.") if w.months >= 12 else (
            80.0, f"Garantia nacional de {w.months} meses.")
    if w.kind == "vendedor":
        return 60.0, f"Garantia do vendedor ({w.months} meses)."
    if w.kind == "internacional":
        return 40.0, "Garantia apenas internacional (pode exigir envio ao exterior)."
    if w.kind == "sem_garantia":
        return 15.0, "Sem garantia informada pelo vendedor."
    return 30.0, "Garantia não informada."


def _conditions_score(offer: Offer, breakdown: PriceBreakdown) -> tuple[float, str]:
    score = 40.0
    parts: list[str] = []
    if breakdown.pix_discount > 0:
        score += 15
        parts.append("desconto no Pix")
    if offer.installments_count >= 6 and offer.installments_interest_free:
        score += 15
        parts.append(f"{offer.installments_count}x sem juros")
    if breakdown.coupon_discount > 0:
        score += 10
        parts.append("cupom validado")
    if breakdown.cashback_later > 0:
        score += 10
        parts.append("cashback posterior")
    if offer.stock is None or offer.stock > 0:
        score += 10
    else:
        parts.append("sem estoque confirmado")
    return min(score, 100.0), ("Condições: " + ", ".join(parts) + ".") if parts else "Condições básicas."


def score_offer(
    ranked: RankedOffer, best_total: float, total_desirable: int
) -> RankedOffer:
    weights = get_settings().ranking_weights
    offer = ranked.offer

    components = {
        "preco_total": _price_score(ranked.price_breakdown.total_delivered, best_total),
        "avaliacoes": _reviews_score(offer),
        "reputacao": _reputation_score(offer.store_reputation, offer.seller_reputation),
        "especificacoes": _specs_score(ranked.desirable_met, total_desirable),
        "historico": _history_score(ranked.history),
        "garantia": _warranty_score(offer),
        "condicoes": _conditions_score(offer, ranked.price_breakdown),
    }
    ranked.score_breakdown = {k: v[0] for k, v in components.items()}
    ranked.score_explanations = {k: v[1] for k, v in components.items()}
    ranked.score = round(sum(weights[k] * v[0] for k, v in components.items()), 1)
    return ranked


def build_alerts(ranked: RankedOffer) -> None:
    """Alertas obrigatorios das secoes 13-18."""
    offer = ranked.offer
    alerts = ranked.alerts

    rep = offer.store_reputation.classification
    if rep in ("ruim", "critica"):
        alerts.append(
            "Esta empresa apresenta preço competitivo, mas possui reputação "
            f"{_REPUTATION_LABEL[rep]} e histórico de reclamações. Compre com cautela."
        )
    if rep == "nao_localizada":
        alerts.append(
            "Não foi localizada reputação suficiente. A avaliação foi baseada em outras evidências disponíveis."
        )
    if offer.warranty.kind == "internacional":
        alerts.append(
            "Este produto possui garantia internacional. Poderá ser necessário enviá-lo ao exterior para reparo."
        )
    if offer.origin == "importado":
        if offer.import_cost and offer.import_cost.risk_of_extra_charges:
            alerts.append("O valor final poderá sofrer cobranças adicionais durante o processo de importação.")
        else:
            alerts.append("Produto importado: confira impostos e prazos antes de comprar.")
    if offer.shipping_days is None:
        alerts.append("Prazo de entrega não informado pela loja.")
    elif offer.shipping_days > 30:
        alerts.append(f"Prazo de entrega longo: {offer.shipping_days} dias.")
    if offer.coupon and not offer.coupon.validated and offer.coupon.code:
        alerts.append(
            f"Cupom '{offer.coupon.code}' disponível, mas não foi possível validar as condições; "
            "não foi descontado do preço total."
        )
    if not offer.installments_interest_free and offer.installments_count > 0:
        alerts.append("O parcelamento desta oferta tem juros e não entrou no cálculo principal.")


def build_pros_cons(ranked: RankedOffer, best_total: float) -> None:
    offer = ranked.offer
    b = ranked.price_breakdown
    if b.total_delivered == best_total:
        ranked.advantages.append("Menor preço total entregue.")
    if offer.reviews.average >= 4.5 and offer.reviews.count >= 100:
        ranked.advantages.append(f"Muito bem avaliado ({offer.reviews.average:.1f}/5).")
    if offer.store_reputation.classification == "excelente":
        ranked.advantages.append("Loja com reputação excelente.")
    if offer.warranty.kind == "nacional" and offer.warranty.months >= 12:
        ranked.advantages.append(f"Garantia nacional de {offer.warranty.months} meses.")
    if b.cashback_later > 0:
        ranked.advantages.append(f"Cashback posterior estimado de R$ {b.cashback_later:.2f}.")
    if offer.installments_count >= 10 and offer.installments_interest_free:
        ranked.advantages.append(f"{offer.installments_count}x sem juros.")

    if offer.reviews.count < 20:
        ranked.disadvantages.append("Poucas avaliações para uma conclusão confiável.")
    if offer.store_reputation.classification in ("regular", "ruim", "critica"):
        ranked.disadvantages.append(
            f"Reputação da loja: {_REPUTATION_LABEL[offer.store_reputation.classification]}."
        )
    if b.shipping > 0:
        ranked.disadvantages.append(f"Frete de R$ {b.shipping:.2f} incluído no total.")
    if offer.reviews.recurring_issues:
        ranked.disadvantages.append("Problemas recorrentes: " + "; ".join(offer.reviews.recurring_issues[:2]) + ".")


def assign_labels(offers: list[RankedOffer]) -> tuple[dict[str, str], dict[str, list[float]]]:
    """Atribui as classificacoes da secao 20 e as faixas de preco da secao 21."""
    highlights: dict[str, str] = {}
    bands: dict[str, list[float]] = {}
    if not offers:
        return highlights, bands

    def tag(label: str, ranked: RankedOffer | None) -> None:
        if ranked is not None:
            highlights[label] = ranked.offer.offer_id
            ranked.labels.append(label)

    by_score = sorted(offers, key=lambda r: r.score, reverse=True)
    by_price = sorted(offers, key=lambda r: r.price_breakdown.total_delivered)

    tag("melhor_custo_beneficio", by_score[0])
    tag("menor_preco", by_price[0])

    reliable = [
        r for r in by_price
        if r.offer.store_reputation.classification in ("excelente", "boa")
        and r.offer.seller_reputation.classification not in ("ruim", "critica")
        and r.offer.reviews.count >= 20
    ]
    tag("menor_preco_confiavel", reliable[0] if reliable else None)

    rated = [r for r in offers if r.offer.reviews.count >= 20]
    tag("melhor_avaliado",
        max(rated, key=lambda r: (r.offer.reviews.average, r.offer.reviews.count)) if rated else None)

    def safety(r: RankedOffer) -> float:
        return (
            _REPUTATION_SCORE[r.offer.store_reputation.classification]
            + (30 if r.offer.warranty.kind == "nacional" else 0)
            + min(r.offer.reviews.count, 1000) / 100
        )
    tag("compra_mais_segura", max(offers, key=safety))

    imported = [r for r in by_score if r.offer.origin == "importado"]
    tag("melhor_importada", imported[0] if imported else None)

    # faixas de preco por distribuicao (tercis), somente com 3+ ofertas
    if len(offers) >= 3:
        prices = sorted(r.price_breakdown.total_delivered for r in offers)

        def tercile(frac: float) -> float:
            k = (len(prices) - 1) * frac
            lo, hi = int(k), min(int(k) + 1, len(prices) - 1)
            return prices[lo] + (prices[hi] - prices[lo]) * (k - lo)

        t1 = tercile(1 / 3)
        t2 = tercile(2 / 3)
        bands = {
            "economica": [prices[0], t1],
            "intermediaria": [t1, t2],
            "premium": [t2, prices[-1]],
        }
        for band, (lo, hi) in bands.items():
            candidates = [r for r in by_score if lo <= r.price_breakdown.total_delivered <= hi]
            tag(f"opcao_{band}", candidates[0] if candidates else None)
        for r in offers:
            p = r.price_breakdown.total_delivered
            r.price_band = "economica" if p <= t1 else "intermediaria" if p <= t2 else "premium"

    return highlights, bands

File: frontend/src/lib/demo/scorer.ts
// Pontuacao 0-100, alertas e classificacoes (porta de backend/app/ranking/scorer.py).
import type { HistoryStats, Offer, PriceBreakdown, RankedOffer } from "../types";

const WEIGHTS: Record<string, number> = {
  preco_total: 0.3,
  avaliacoes: 0.2,
  reputacao: 0.15,
  especificacoes: 0.15,
  historico: 0.1,
  garantia: 0.05,
  condicoes: 0.05,
};

const REP_SCORE: Record<string, number> = {
  excelente: 100, boa: 80, regular: 55, ruim: 25, critica: 10, insuficiente: 40, nao_localizada: 40,
};
const HIST_SCORE: Record<string, number> = {
  muito_baixo: 100, baixo: 80, na_media: 60, alto: 30, muito_alto: 10, insuficiente: 50,
};
const REP_LABEL: Record<string, string> = {
  excelente: "excelente", boa: "boa", regular: "regular", ruim: "ruim", critica: "crítica",
  insuficiente: "insuficiente", nao_localizada: "não localizada",
};

type Component = [number, string];

function priceScore(total: number, best: number): Component {
  if (total <= 0 || best <= 0) return [0, "Preço total indisponível."];
  const score = Math.round(Math.min(100, (best / total) * 100) * 10) / 10;
  if (total === best) return [score, "Menor preço total entregue entre as ofertas válidas."];
  return [score, `Preço total ${((total / best - 1) * 100).toFixed(0)}% acima da oferta mais barata válida.`];
}

function reviewsScore(offer: Offer): Component {
  const r = offer.reviews;
  if (r.count === 0) return [30, "Sem avaliações localizadas; nota neutra baixa."];
  const confidence = Math.min(1, Math.log10(r.count + 1) / 3);
  const score = Math.round((r.average / 5) * 100 * (0.5 + 0.5 * confidence) * 10) / 10;
  const level = confidence > 0.8 ? "alta" : confidence > 0.5 ? "média" : "baixa";
  return [score, `Nota ${r.average.toFixed(1)}/5 em ${r.count} avaliações (confiança ${level}).`];
}

function reputationScore(offer: Offer): Component {
  const store = offer.store_reputation.classification;
  const seller = offer.seller_reputation.classification;
  const sStore = REP_SCORE[store] ?? 40;
  if (seller === "nao_localizada" || seller === "insuficiente")
    return [sStore, `Loja com reputação ${REP_LABEL[store] ?? store}.`];
  const score = Math.round((sStore * 0.6 + (REP_SCORE[seller] ?? 40) * 0.4) * 10) / 10;
  return [score, `Loja ${REP_LABEL[store] ?? store}, vendedor ${REP_LABEL[seller] ?? seller}.`];
}

function specsScore(desirableMet: string[], totalDesirable: number): Component {
  if (totalDesirable === 0) return [60, "Nenhum critério desejável informado; nota neutra."];
  const score = Math.round((40 + (desirableMet.length / totalDesirable) * 60) * 10) / 10;
  return [score, `Atende ${desirableMet.length} de ${totalDesirable} critérios desejáveis.`];
}

function historyScore(h: HistoryStats): Component {
  const score = HIST_SCORE[h.classification] ?? 50;
  if (h.classification === "insuficiente") return [score, "Histórico de preços insuficiente; nota neutra."];
  return [score, `Preço atual classificado como '${h.classification.replace(/_/g, " ")}' frente ao histórico.`];
}

function warrantyScore(offer: Offer): Component {
  const w = offer.warranty;
  if (w.kind === "nacional")
    return w.months >= 12
      ? [100, `Garantia nacional de ${w.months} meses.`]
      : [80, `Garantia nacional de ${w.months} meses.`];
  if (w.kind === "vendedor") return [60, `Garantia do vendedor (${w.months} meses).`];
  if (w.kind === "internacional") return [40, "Garantia apenas internacional (pode exigir envio ao exterior)."];
  if (w.kind === "sem_garantia") return [15, "Sem garantia informada pelo vendedor."];
  return [30, "Garantia não informada."];
}

function conditionsScore(offer: Offer, b: PriceBreakdown): Component {
  let score = 40;
  const parts: string[] = [];
  if (b.pix_discount > 0) { score += 15; parts.push("desconto no Pix"); }
  if (offer.installments_count >= 6 && offer.installments_interest_free) {
    score += 15;
    parts.push(`${offer.installments_count}x sem juros`);
  }
  if (b.coupon_discount > 0) { score += 10; parts.push("cupom validado"); }
  if (b.cashback_later > 0) { score += 10; parts.push("cashback posterior"); }
  const stock = (offer as Offer & { stock?: number | null }).stock;
  if (stock === undefined || stock === null || stock > 0) score += 10;
  else parts.push("sem estoque confirmado");
  return [Math.min(score, 100), parts.length ? `Condições: ${parts.join(", ")}.` : "Condições básicas."];
}

export function scoreOffer(ranked: RankedOffer, bestTotal: number, totalDesirable: number): void {
  const components: Record<string, Component> = {
    preco_total: priceScore(ranked.price_breakdown.total_delivered, bestTotal),
    avaliacoes: reviewsScore(ranked.offer),
    reputacao: reputationScore(ranked.offer),
    especificacoes: specsScore(ranked.desirable_met, totalDesirable),
    historico: historyScore(ranked.history),
    garantia: warrantyScore(ranked.offer),
    condicoes: conditionsScore(ranked.offer, ranked.price_breakdown),
  };
  ranked.score_breakdown = Object.fromEntries(Object.entries(components).map(([k, v]) => [k, v[0]]));
  ranked.score_explanations = Object.fromEntries(Object.entries(components).map(([k, v]) => [k, v[1]]));
  ranked.score =
    Math.round(
      Object.entries(components).reduce((sum, [k, v]) => sum + WEIGHTS[k] * v[0], 0) * 10,
    ) / 10;
}

export function buildAlerts(ranked: RankedOffer): void {
  const offer = ranked.offer;
  const alerts = ranked.alerts;
  const rep = offer.store_reputation.classification;
  if (rep === "ruim" || rep === "critica")
    alerts.push(
      `Esta empresa apresenta preço competitivo, mas possui reputação ${REP_LABEL[rep]} e histórico de reclamações. Compre com cautela.`,
    );
  const sellerRep = offer.seller_reputation.classification;
  if ((sellerRep === "ruim" || sellerRep === "critica") && rep !== "ruim" && rep !== "critica")
    alerts.push(
      `O vendedor ${offer.seller_name || "do marketplace"} possui reputação ${REP_LABEL[sellerRep]} e elevado índice de reclamações não solucionadas.`,
    );
  if (rep === "nao_localizada")
    alerts.push(
      "Não foi localizada reputação suficiente. A avaliação foi baseada em outras evidências disponíveis.",
    );
  if (offer.warranty.kind === "internacional")
    alerts.push(
      "Este produto possui garantia internacional. Poderá ser necessário enviá-lo ao exterior para reparo.",
    );
  if (offer.origin === "importado") {
    const importCost = (offer as Offer & { import_cost?: { risk_of_extra_charges?: boolean } | null }).import_cost;
    if (importCost?.risk_of_extra_charges)
      alerts.push("O valor final poderá sofrer cobranças adicionais durante o processo de importação.");
    else alerts.push("Produto importado: confira impostos e prazos antes de comprar.");
  }
  if (offer.shipping_days === null || offer.shipping_days === undefined)
    alerts.push("Prazo de entrega não informado pela loja.");
  else if (offer.shipping_days > 30) alerts.push(`Prazo de entrega longo: ${offer.shipping_days} dias.`);
  if (offer.coupon && !offer.coupon.validated && offer.coupon.code)
    alerts.push(
      `Cupom '${offer.coupon.code}' disponível, mas não foi possível validar as condições; não foi descontado do preço total.`,
    );
  if (!offer.installments_interest_free && offer.installments_count > 0)
    alerts.push("O parcelamento desta oferta tem juros e não entrou no cálculo principal.");
}

export function buildProsCons(ranked: RankedOffer, bestTotal: number): void {
  const offer = ranked.offer;
  const b = ranked.price_breakdown;
  if (b.total_delivered === bestTotal) ranked.advantages.push("Menor preço total entregue.");
  if (offer.reviews.average >= 4.5 && offer.reviews.count >= 100)
    ranked.advantages.push(`Muito bem avaliado (${offer.reviews.average.toFixed(1)}/5).`);
  if (offer.store_reputation.classification === "excelente")
    ranked.advantages.push("Loja com reputação excelente.");
  if (offer.warranty.kind === "nacional" && offer.warranty.months >= 12)
    ranked.advantages.push(`Garantia nacional de ${offer.warranty.months} meses.`);
  if (b.cashback_later > 0)
    ranked.advantages.push(`Cashback posterior estimado de R$ ${b.cashback_later.toFixed(2)}.`);
  if (offer.installments_count >= 10 && offer.installments_interest_free)
    ranked.advantages.push(`${offer.installments_count}x sem juros.`);

  if (offer.reviews.count < 20) ranked.disadvantages.push("Poucas avaliações para uma conclusão confiável.");
  const rep = offer.store_reputation.classification;
  if (rep === "regular" || rep === "ruim" || rep === "critica")
    ranked.disadvantages.push(`Reputação da loja: ${REP_LABEL[rep]}.`);
  if (b.shipping > 0) ranked.disadvantages.push(`Frete de R$ ${b.shipping.toFixed(2)} incluído no total.`);
  if (offer.reviews.recurring_issues.length > 0)
    ranked.disadvantages.push(
      `Problemas recorrentes: ${offer.reviews.recurring_issues.slice(0, 2).join("; ")}.`,
    );
}

export function assignLabels(
  offers: RankedOffer[],
): [Record<string, string>, Record<string, number[]>] {
  const highlights: Record<string, string> = {};
  const bands: Record<string, number[]> = {};
  if (offers.length === 0) return [highlights, bands];

  const tag = (label: string, r: RankedOffer | undefined | null) => {
    if (r) {
      highlights[label] = r.offer.offer_id;
      r.labels.push(label);
    }
  };

  const byScore = [...offers].sort((a, b) => b.score - a.score);
  const byPrice = [...offers].sort(
    (a, b) => a.price_breakdown.total_delivered - b.price_breakdown.total_delivered,
  );

  tag("melhor_custo_beneficio", byScore[0]);
  tag("menor_preco", byPrice[0]);

  const reliable = byPrice.filter(
    (r) =>
      ["excelente", "boa"].includes(r.offer.store_reputation.classification) &&
      !["ruim", "critica"].includes(r.offer.seller_reputation.classification) &&
      r.offer.reviews.count >= 20,
  );
  tag("menor_preco_confiavel", reliable[0]);

  const rated = offers.filter((r) => r.offer.reviews.count >= 20);
  tag(
    "melhor_avaliado",
    rated.length
      ? rated.reduce((best, r) =>
          r.offer.reviews.average > best.offer.reviews.average ||
          (r.offer.reviews.average === best.offer.reviews.average &&
            r.offer.reviews.count > best.offer.reviews.count)
            ? r
            : best,
        )
      : null,
  );

  const safety = (r: RankedOffer) =>
    (REP_SCORE[r.offer.store_reputation.classification] ?? 40) +
    (r.offer.warranty.kind === "nacional" ? 30 : 0) +
    Math.min(r.offer.reviews.count, 1000) / 100;
  tag("compra_mais_segura", offers.reduce((best, r) => (safety(r) > safety(best) ? r : best)));

  const imported = byScore.filter((r) => r.offer.origin === "importado");
  tag("melhor_importada", imported[0]);

  if (offers.length >= 3) {
    const prices = offers.map((r) => r.price_breakdown.total_delivered).sort((a, b) => a - b);
    const tercile = (frac: number) => {
      const k = (prices.length - 1) * frac;
      const lo = Math.floor(k);
      const hi = Math.min(lo + 1, prices.length - 1);
      return prices[lo] + (prices[hi] - prices[lo]) * (k - lo);
    };
    const t1 = tercile(1 / 3);
    const t2 = tercile(2 / 3);
    bands["economica"] = [prices[0], t1];
    bands["intermediaria"] = [t1, t2];
    bands["premium"] = [t2, prices[prices.length - 1]];
    for (const [band, [lo, hi]] of Object.entries(bands)) {
      const candidates = byScore.filter(
        (r) => r.price_breakdown.total_delivered >= lo && r.price_breakdown.total_delivered <= hi,
      );
      tag(`opcao_${band}`, candidates[0]);
    }
    for (const r of offers) {
      const p = r.price_breakdown.total_delivered;
      r.price_band = p <= t1 ? "economica" : p <= t2 ? "intermediaria" : "premium";
    }
  }

  return [highlights, bands];
}
